File: examples/ios-sdk-examples/ios-sdk-examples-objc/Views/TileOverlay/TileOverlayViewController.m
//
//  TileOverlayViewController.m
//  ios-sdk-examples-objc
//
//  Created by Huy Dang on 25/01/2021.
//

#import "TileOverlayViewController.h"
#import <Map4dMap/Map4dMap.h>
#import "TileURLConstructor.h"

@interface TileOverlayViewController()

@property (weak, nonatomic) IBOutlet MFMapView *mapView;

@property (strong, nonatomic, nullable) MFURLTileLayer *tileOverlay;

@end

@implementation TileOverlayViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  [self createTileOverlay];
  [self addTileOverlay];
}

- (void) createTileOverlay {
  TileURLConstructor * urlConstructor = [[TileURLConstructor alloc] init];
  MFURLTileLayer * tileOverlay = [MFURLTileLayer tileLayerWithURLConstructor:urlConstructor];
//  [tileOverlay setMap:self.mapView];
  self.tileOverlay = tileOverlay;
}

- (void) hideTileOverlay {
  self.tileOverlay.isHidden = true;
}

- (void) showTileOverlay {
  self.tileOverlay.isHidden = false;
}

- (void) addTileOverlay {
  self.tileOverlay.map = self.mapView;
}

- (void) removeTileOverlay {
  self.tileOverlay = nil;
}

@end

File: examples/ios-sdk-examples/ios-sdk-examples-swift/Views/TileOverlay/TileOverlayViewController.swift
import UIKit
import Map4dMap

class TileOverlayViewController : UIViewController {
  
  @IBOutlet weak var addRemove: UIButton!
  @IBOutlet weak var hideShow: UIButton!
  @IBOutlet weak var mapView: MFMapView!
  
  private lazy var tileOverlay: MFURLTileLayer = {
    let urlConstructor = TileURLConstructor()
    let tileOverlay = MFURLTileLayer(urlConstructor: urlConstructor)
    return tileOverlay
  }()
  
  override func viewWillAppear(_ animated: Bool) {
    super.viewWillAppear(animated)
    navigationItem.title = "Tile Overlay"
  }
  
  private func addTileOverlay() {
    tileOverlay.map = mapView
    
    showTileOverlay()
    addRemove.setTitle("Remove TileOverlay", for: .normal)
    hideShow.isHidden = false
  }
  
  private func removeTileOverlay() {
    tileOverlay.map = nil
    
    addRemove.setTitle("Add TileOverlay", for: .normal)
    hideShow.isHidden = true
  }
  
  private func hideTileOverlay() {
    tileOverlay.isHidden = true
    
    hideShow.setTitle("Show TileOverlay", for: .normal)
  }
  
  private func showTileOverlay() {
    tileOverlay.isHidden = false
    
    hideShow.setTitle("Hide TileOverlay", for: .normal)
  }
  
  @IBAction func didTapAddRemove(_ sender: Any) {
    if tileOverlay.map != nil {
      removeTileOverlay()
    }
    else {
      addTileOverlay()
    }
  }
  
  @IBAction func didTapShowHide(_ sender: Any) {
    if (tileOverlay.isHidden) {
      showTileOverlay()
    }
    else {
      hideTileOverlay()
    }
  }
  
}

extension TileOverlayViewController {
  class TileURLConstructor : NSObject, MFTileURLConstructor {
    func getTileUrlWith(x: UInt, y: UInt, zoom: UInt, is3dMode: Bool) -> URL? {
      let url = "https://tile.openstreetmap.de/\(zoom)/\(x)/\(y).png"
      return URL(string: url)
    }
  }
}
